### PR TITLE
buildcache: skip unrecognized metadata files

### PR DESCRIPTION
**Pull request by scottwittenburg**
This PR ensures that unrecognized metadata files do not cause download failure, but are skipped.

This improves forward compatibility with future metadata formats.

### CODE REVIEW

**Comment by scottwittenburg** [comment]
fyi @kwryankrattiger

**Comment by scottwittenburg** [comment]
It's weird, the new test failed on mac, seemingly because all the expected output bits were not found in the output.  Specifically, it seems to think this [line](https://github.com/spack/spack/actions/runs/6805671623/job/18505782564?pr=40941#step:6:108) was missing from the output, and yet there it is.

**Comment by haampie** [comment]
Can the test be just a *fast* unit test of `download_tarball`? Doesn't really need concretization, and you wouldn't have to deal with avoiding side effects you're describing above.

File: lib/spack/spack/binary_distribution.py [hunk 1]
```diff
@@ -1682,6 +1707,15 @@ def download_tarball(spec, unsigned=False, mirrors_for_spec=None):
                     local_specfile_path = local_specfile_stage.save_filename
                     signature_verified = False
 
+                    spec_dict = _read_spec_file(local_specfile_path)
+                    layout_version = int(spec_dict.get("buildcache_layout_version", 0))
```
Comment (haampie) [inline lib/spack/spack/binary_distribution.py]: Deal with ValueError

File: lib/spack/spack/binary_distribution.py [hunk 2]
```diff
@@ -2001,24 +2035,16 @@ def extract_tarball(spec, download_result, unsigned=False, force=False, timer=ti
...
 
     filename = download_result["tarball_stage"].save_filename
     signature_verified = download_result["signature_verified"]
     tmpdir = None
 
-    if (
-        "buildcache_layout_version" not in spec_dict
-        or int(spec_dict["buildcache_layout_version"]) < 1
-    ):
+    layout_version = int(spec_dict.get("buildcache_layout_version", 0))
```
Comment (haampie) [inline lib/spack/spack/binary_distribution.py]: Same

File: lib/spack/spack/binary_distribution.py [hunk 3]
```diff
@@ -1682,6 +1707,15 @@ def download_tarball(spec, unsigned=False, mirrors_for_spec=None):
                     local_specfile_path = local_specfile_stage.save_filename
                     signature_verified = False
 
+                    spec_dict = _read_spec_file(local_specfile_path)
+                    layout_version = int(spec_dict.get("buildcache_layout_version", 0))
+                    if layout_version > _current_build_cache_layout_version:
```
Comment (haampie) [inline lib/spack/spack/binary_distribution.py]: Duplicate this logic in the `parsed.scheme == "oci"` branch

File: lib/spack/spack/binary_distribution.py [hunk 4]
```diff
@@ -1560,6 +1574,17 @@ def _delete_staged_downloads(download_result):
...
 
 
+def _read_spec_file(specfile_path):
+    spec_dict = {}
+    with open(specfile_path, "r") as inputfile:
+        content = inputfile.read()
+        if specfile_path.endswith(".json.sig"):
+            spec_dict = Spec.extract_json_from_clearsig(content)
+        else:
+            spec_dict = sjson.load(content)
```
Comment (haampie) [inline lib/spack/spack/binary_distribution.py]: Deal with this throwing

**Comment by haampie** [comment]
While we're at this, can you also "reserve" the option for (a) transparently compressed json files, and (b) clear-signed spec.json files without *.sig extension? This has to be handled before parsing as json, and currently parsing errors are caught.

At minimum catch any exception about json / clearsign json parsing failures.

But slightly better: read the file in binary mode and parse as normal (clearsigned json or json). On failure check magic bytes for gzip and give smth informative like "compressed metadata files unsupported" or on "magic bytes" `b"-----BEGIN PGP SIGNED MESSAGE-----"` warn that clearsigned `spec.json` files are not supported -- although, it's not very hard to support this also in 0.21.

Thanks

**Comment by haampie** [comment]
@scottwittenburg given the lack of time I took over your PR instead of reviewing it.

File: lib/spack/spack/binary_distribution.py [hunk 5]
```diff
@@ -1560,6 +1565,42 @@ def _delete_staged_downloads(download_result):
...
+        if path.endswith(".json.sig"):
+            spec_dict = Spec.extract_json_from_clearsig(as_string)
+        else:
+            spec_dict = json.loads(as_string)
+    except Exception as e:
+        raise InvalidMetadataFile(f"Could not parse {path} due to: {e}") from e
+
+    # Ensure this version is not too new.
+    try:
+        layout_version = int(spec_dict.get("buildcache_layout_version", 0))
```
Comment (alalazo) [inline lib/spack/spack/binary_distribution.py:1592]: (Not a request, just a comment) Do we have any formal specification (a schema?) of what a spec_dict looks like?

File: lib/spack/spack/binary_distribution.py [hunk 6]
```diff
@@ -1652,6 +1693,18 @@ def download_tarball(spec, unsigned=False, mirrors_for_spec=None):
                     try:
                         local_specfile_stage.fetch()
                         local_specfile_stage.check()
+                        try:
+                            _get_valid_spec_file(
+                                local_specfile_stage.save_filename,
+                                CURRENT_BUILD_CACHE_LAYOUT_VERSION,
+                            )
```
Comment (alalazo) [inline lib/spack/spack/binary_distribution.py:1700]: (Just a comment) Wondering if we can turn this double nested `try/except` into a context manager

File: lib/spack/spack/binary_distribution.py [hunk 7]
```diff
@@ -614,11 +619,11 @@ def compute_hash(data):
 
 
 def build_cache_relative_path():
-    return _build_cache_relative_path
+    return BUILD_CACHE_RELATIVE_PATH
```
Comment (alalazo) [inline lib/spack/spack/binary_distribution.py:622]: (minor) This function and the one below are probably unnecessary

File: lib/spack/spack/binary_distribution.py [hunk 8]
```diff
@@ -1401,7 +1406,7 @@ def _build_tarball_in_stage_dir(spec: Spec, out_url: str, stage_dir: str, option
             spec_dict = sjson.load(content)
         else:
             raise ValueError("{0} not a valid spec file type".format(spec_file))
-    spec_dict["buildcache_layout_version"] = 1
+    spec_dict["buildcache_layout_version"] = CURRENT_BUILD_CACHE_LAYOUT_VERSION
```
Comment (alalazo) [inline lib/spack/spack/binary_distribution.py:1409]: (Minor) We should probably decide whether it's `buildcache` or `build_cache`

File: lib/spack/spack/binary_distribution.py [hunk 9]
```diff
@@ -2029,7 +2086,7 @@ def extract_tarball(spec, download_result, unsigned=False, force=False, timer=ti
             _delete_staged_downloads(download_result)
             shutil.rmtree(tmpdir)
             raise e
-    else:
+    elif layout_version == 1:
```
Comment (alalazo) [inline lib/spack/spack/binary_distribution.py:2089]: Do we need an `else` case here for unknown layouts?

**Review by alalazo — APPROVED** [review]
I didn't test this extensively, but from a first read it seems fine. I just left a few comments that, in case, can be addressed in a later PR.

File: lib/spack/spack/binary_distribution.py (hunk 9 above)
Comment (haampie) [reply lib/spack/spack/binary_distribution.py:2089]: That already throws earlier

File: lib/spack/spack/binary_distribution.py (hunk 5 above)
Comment (haampie) [reply lib/spack/spack/binary_distribution.py:1592]: I believe there was, but pretty sure it's unused or outdated

**Pull request merged**